src/Workspaces/Core/Portable/Utilities/OwnedDisposable`1.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Roslyn.Utilities
+{
+    internal struct OwnedDisposable<T> : IDisposable, IEquatable<OwnedDisposable<T>>
```

Comment:
:memo: This type would need analyzer support to ensure it's treated as non-copyable. It can either be passed by reference or moved to a different instance. For cases where it's used but will not transfer ownership, `in` can be used.

Reply:
some questions. why it is struct instead of class? is that to force having ref?

if it require special analyzer support to detect, how about creating well known interface such as IOwnerTrackingDisposable or something, that analyzer knows about, and has well known operation such as Claims, Move and etc below so that analyzer can track ownership more clearly and detect mis-usage such as assigning disposable object to another without transferring ownership?

is the point having wrapper type that can always disposed within same method?

Reply:
> why it is struct instead of class?

Allocation-free fast path. The type should not incur a performance penalty.

> if it require special analyzer support to detect, how about creating well known interface such as IOwnerTrackingDisposable or something, that analyzer knows about, and has well known operation such as Claims, Move and etc below so that analyzer can track ownership more clearly and detect mis-usage such as assigning disposable object to another without transferring ownership?

This is intended to be that interface. The use of `struct` vs `interface` for it goes back to the previous item.

> is the point having wrapper type that can always disposed within same method?

Yes, this is a requirement of the current `IDisposable` analyzer design.